File: leetcode/078_subsets.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def backtrack(result, nums, currlist, start):
            result.append(currlist)
            for i in range(start, len(nums)):
                backtrack(result, nums, currlist + [nums[i]], i+1)
        res = []
        backtrack(res, nums, [], 0)
        return res
        

File: leetcode/078_subsets.ts
function subsets(nums: number[]): number[][] {
    const res = []
    generate(nums, [], 0, res)
    return res
};

function generate(nums: number[], currSet: number[], idx: number, result: number[][]) {
    result.push(currSet)
    if (idx >= nums.length) {
        return
    }
    for (let i = idx; i < nums.length; i++) {
        generate(nums, [...currSet, nums[i]], i+1, result)
    }  
}
